Implement HtmlUtils helper functions

andy-armstrong opened the pull request:
https://openedx.atlassian.net/browse/UITK-74

This PR adds a new HtmlUtils class to the UI Toolkit, with two functions:
 - ``HTML(text)`` - wraps a known piece of HTML so that it doesn't get double escaped
 - ``interpolate(string, options)`` - returns an HTML-escaped string where the interpolation escapes each option unless it is an instance of an HTMLSnippet (the result of calling HTML)

Note that we have gone around and around on the name of the helper function. I think it is important that it include the word ``escape`` as its behavior is to return an HTML-escaped string, so I ended up going with Robert's original suggestion.

Once this version of the UI Toolkit has been released, I'll make another PR against edx-platform which will bring in this function and replace some or all usages of ``StringUtils.interpolate_text`` with this new, safer function. We will then update the XSS guidelines to advise folks to use this function instead of any of the alternatives, i.e. interpolate, interpolate_text and _.template.

@AlasdairSwan @robrap please review.

FYI @dan-f @nedbat @cahrens

edx-webhook commented:
## PR Cover Letter
- [ ] A detailed description of the changes in the body of the PR
- [ ] If available, a link to the relevant JIRA ticket in the body of the PR
  - `[PROJ-XXXX](https://openedx.atlassian.net/browse/PROJ-XXXX)`

## Sandbox
- [ ] Build a sandbox for your branch and add a link here

## Testing
- [ ] i18n
- [ ] RTL
- [ ] Unit, integration, acceptance tests as appropriate
- [ ] Analytics
- [ ] Performance
- [ ] Database migrations are backwards-compatible

## Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [ ] Code review #1
- [ ] Code review #2
- [ ] Docs review (required for UI strings/error messages)
- [ ] UX review
- [ ] Accessibility review
- [ ] Product review

## Documentation considerations
- [ ] Are there user visible changes? If so, list who is affected, and include a pointer to the relevant code.
- [ ] Should this be mentioned in the release notes? If so, make sure that you included a description.

## DevOps assistance
If it's needed, or even if you suspect it might be needed, do any/all these things as early as possible. Don't surprise devops right before merging! For example, touching envs/aws.py or or envs/common.py are dead giveaways that you'll need assistance when pushing your changes out. Things that require devops attention include, but are not limited to:
- configuration changes
- feature flags
- migrations
Ways to get devops attention, pick-and-choose whichever is most appropriate for your specific PR:
- [ ] Tag devops on your PR (`@edx/devops`), either for review (more specific to devops = better) or just a general FYI
- [ ] File a support ticket

FYI: Tag anyone who might be interested in this PR here.

## Post-review
- [ ] Squash commits

AlasdairSwan commented on src/js/utils/string-utils.js:
```diff
@@ -0,0 +1,91 @@
+     * @param {string} text The HTML for the text.
+     * @constructor
+     */
+    function HTMLSnippet(text) {
```
This function hasn't been defined as a var.

AlasdairSwan commented on src/js/utils/string-utils.js:
```diff
@@ -0,0 +1,91 @@
+        var escapedOptions = escapeInterpolateOptions(options),
+            result = _.escape(string).replace(/{\w+}/g,
+                function(parameter) {
+                    return String(escapedOptions[parameter.slice(1,-1)]);
```
Can you explain what the negative end slice is doing here?

robrap replied:
I think these are just dropping the `{` and `}` that are wrapping the parameter name.

AlasdairSwan replied:
So ```-1``` is basically ```length-1```

robrap replied:
Yes.  That's not as clear as it should be in [this documentation](http://www.w3schools.com/jsref/jsref_slice_string.asp), but you can see it as a "Tip" and the final example.

robrap commented on src/js/utils/specs/string-utils-spec.js:
```diff
@@ -0,0 +1,58 @@
+        describe('StringUtils', function() {
+            describe('interpolateEscape', function() {
+                it('can interpolate a string with no parameters provided', function() {
+                    expect(StringUtils.interpolateEscape('Hello, world').toString()).toBe(
```
I think this test can be removed, because it duplicates the first withData test.

andy-armstrong replied:
It is subtly different because it tests the case where the options parameter isn't even provided. I tried to think of a way to incorporate this into the other tests but it seemed simpler to keep it separate.

robrap replied:
I see.  Then maybe below you can change the comment to "...with empty parameters".

robrap commented on src/js/utils/string-utils.js:
```diff
@@ -0,0 +1,91 @@
+    };
+
+    /**
+     * Returns an HTML-escaped text string by substituting values into
```
Maybe start with (or include) the following:
Returns an HTMLSnippet. ...

andy-armstrong replied:
The way I have this right now, HTMLSnippet is an internal class, so I didn't want to document it. I suppose I could make it public and then the function would be typed more appropriately.

robrap replied:
I think HTMLSnippet needs to be external.  

We should agree on its name as well.  As you can see in my comment on the "[Implement in the platform ticket](https://openedx.atlassian.net/browse/UITK-75?focusedCommentId=176420&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-176420)", code outside of this method may want to check its type (String or HTMLSnippet), and determine whether or not it still needs to escape.

robrap commented on src/js/utils/string-utils.js:
```diff
@@ -0,0 +1,91 @@
+     *
+     * Example:
+     *   StringUtils.interpolateEscape(
+     *       'You are enrolling in {spanStart}{courseName}{spanEnd}',
```
Could you add the gettext() call into this example code.  If you find it necessary, you could add a note that `gettext()` is not strictly needed for interpolateEscape() to work, but it is a requirement for following the best practice of translating strings properly.  I'd just hate to have people copy and paste a code-snippet that doesn't do the right thing.

andy-armstrong replied:
As we discussed, gettext is a Django i18n function, so it is a little strange to embed that into a general purpose JavaScript library that might be used in IDAs that are not Django based. I'm going to leave the example as is, but then add a description about the needs of i18n.

robrap commented:
@andy-armstrong This looks great.  Only some very minor comments.

dan-f commented on src/js/utils/string-utils.js:
```diff
@@ -0,0 +1,91 @@
+        var property, value,
+            result = {};
+        for (property in options) {
+            if (hasOwnProperty.call(options, property)) {
```
Using `_.mapObject` might make this a bit more readable as it only iterates over an object's own properties, and the iteratee callback gets the key/value pair (no need to index into the object). It also returns a new object which removes the need to manually manage the `result` object.

andy-armstrong replied:
That is so much easier! I didn't know about this function before (and amazingly there are no usages of it in any of the platform's code that I can see).

dan-f commented:
I know you all are probably sick of hearing me push back against names, but I think they're important, especially for a core API like this. I'm still not on board with `interpolateEscape` because I feel that it highlights an implementation detail over the actual use case. From a developer (user) perspective, I want to interpolate some HTML strings, and the escaping should Just Work™.

From this perspective, I'd like to propose `interpolateHtml` as the new name for this method because that's what we're actually doing; interpolating an HTML string, where escaping - an implementation detail - is handled by the implementation of the API.

andy-armstrong commented:
@dan-f Naming is both important and hard so I appreciate your perseverance. :-) I initially proposed ``interpolateHtml`` yesterday when I started this, but @robrap convinced me that other JS functions don't have HTML in the name e.g. ``_.escape`` is HTML escaping, but it is focused on the action not the content type. 

One aspect to this is that ``interpolateHtml`` reads like an action, while ``interpolateEscape`` isn't really even a phrase. The former seems a more pleasing name, which must count for something. :-) In Java, the basic escape function is called [escapeHTML](https://commons.apache.org/proper/commons-lang/javadocs/api-2.6/org/apache/commons/lang/StringEscapeUtils.html#escapeHtml(java.lang.String)) which does seem clearer, because you can escape URLs or other formats too.

All this is to say that I think I'm now with you @dan-f . @robrap @AlasdairSwan @nedbat do you have strong preferences either way? Any other name that you think would work better?

dan-f commented:
Thanks for the thoughtful response, @andy-armstrong! +1 on `interpolateEscape` not being a phrase.

I see where you're coming from on the whole not calling out the content type, but I also like the idea of eventually having interpolation functions for URLs; we could have a `StringUtils.interpolateURI` which has a similar API but uses a `URI` constructor and wraps up `encodeURIComponent`.

andy-armstrong commented:
@dan-f After trying this refactoring, I'm now tempted to rename the file to ``html-utils.js`` since all of the functions are to do with managing HTML. It would then look like ``HtmlUtils.interpolateHtml`` and ``HtmlUtils.HTML('<a href="#"></a>')``. Thoughts?

andy-armstrong commented:
@dan-f pointed out to me that if the class is HtmlUtils then we don't need the suffix in the name. So how about ``HtmlUtils.interpolate``? The name seems unambiguous that this function is the correct interpolation function to use when handling HTML.

dan-f commented:
I like `HtmlUtils.interpolate` & `HtmlUtils.Html`.

robrap commented:
I'm fine with `HtmlUtils.interpolate` & `HtmlUtils.HTML`.  @dan-f We intentionally decided to go with all-caps for the `HTML` method because we liked how it looked better, but it may be different here if it is following `HtmlUtils`. @nedbat Thoughts?
Also, is everyone fine with `HtmlUtils.HtmlSnippet`, or are there opinions on that as well?

dan-f commented:
I prefer `Html` over `HTML` since 1) we're distinguishing between JS name, not a general acronym, and 2) it's easier to type. In fact, I now wonder if `HtmlUtils.Html` should actually be `HtmlUtils.html` since it's a factory function rather than a constructor.

I don't mind `HtmlUtils.HtmlSnippet` though I also like `HtmlUtils.HtmlString` as it's a bit shorter and a little bit more specific. I don't mind the distinction here though.

Sorry for being so much trouble! :)

dan-f commented:
OK I have another idea. Sorry folks. Really! Sorry for the churn!

I want to challenge whether we need both a `HTML` factory function and an `interpolateHtml` function. Instead, I'll propose a polymorphic function named `Html`. With a single argument, it returns that string (just like the `HTML` factory currently does). With a second interpolation argument, it does the interpolation:

```javascript
var Html = require('html-utils').Html;

Html('Please click on {linkStart}this awesome link{linkEnd}.', {
    linkStart: Html('<a href="/the-awesome-page/">'),
    linkEnd: Html('<a href="/the-awesome-page/">')
})

Html('Also check out not-prone we are to XSS: {htmlText}', {
    htmlText: '<script>alert("This will not open an alert.")</script>'
})
```

robrap commented:
I think we need 2 calls.  It does not make sense to me that this would escape the ``&``:
```
Html('Please click on {linkStart}this great & awesome link{linkEnd}.', {
    linkStart: Html('<a href="/the-awesome-page/">'),
    linkEnd: Html('</a>')
```
But, this would not escape the ``&``:
```
Html('This is a great & awesome example.')
``` 

Also, we went for capitalization because we wanted it to be more declarative.  It just feel more right to me to use `Html('<a>')` than `html('<a>') in this case, even though we are forcing the capitalization a bit.

andy-armstrong commented:
@robrap @dan-f @AlasdairSwan Thanks for the reviews. I've pushed up a commit which addresses most of the feedback. We will need to chat more on Monday to settle upon the name of the HTML factory method. I've currently left it as it was.

dan-f commented on src/js/utils/html-utils.js:
```diff
@@ -0,0 +1,115 @@
+         *   HtmlUtils.interpolate(
+         *       'You are enrolling in {spanStart}{courseName}{spanEnd}',
+         *       {
+     *           courseName: 'Rock & Roll 101',
```
Looks like these lines got accidentally de-indented.

andy-armstrong replied:
Whoops! I've fixed it.

dan-f commented on src/js/utils/html-utils.js:
```diff
@@ -0,0 +1,115 @@
+        return {
+            interpolate: interpolate,
+            HTML: HTML,
+            HtmlSnippet: HtmlSnippet
```
What's the use case for this being public?

andy-armstrong replied:
See @robrap's comment earlier in this PR:

{quote}As you can see in my comment on the "Implement in the platform ticket" ([UITK-75](https://openedx.atlassian.net/browse/UITK-75)), code outside of this method may want to check its type (String or HTMLSnippet), and determine whether or not it still needs to escape.{quote}

It also makes it somewhat clearer to document the helper methods as they can say what type they return.

dan-f replied:
If that's a common use case, then I think perhaps we should add it to this library.

andy-armstrong replied:
Good point, @dan-f. I think my new method ``HtmlUtils.escape`` does exactly what @robrap wanted, in that it will only escape if necessary. I suppose you could now argue that that means that ``HtmlSnippet`` can become internal again. I'm tempted to leave it public so that the documentation is clearer, and because I could imagine there being other potential usages.

dan-f commented:
:+1: once we figure out names!

robrap commented on src/js/utils/html-utils.js:
```diff
@@ -0,0 +1,115 @@
+         *
+         * The intention of an HTML snippet is to capture a string that
+         * is known to contain HTML that has been safely escaped.
+         * This allows interpolate to understand that it does not
```
Possible adjustment of "This allows interpolate to..." to "As an example, this allows interpolate to...".  This might clarify that other outside code could have the same understanding as interpolate.

andy-armstrong replied:
Fixed, thanks

robrap commented on src/js/utils/specs/html-utils-spec.js:
```diff
@@ -0,0 +1,70 @@
+    function(SpecHelpers, HtmlUtils) {
+        'use strict';
+
+        describe('HTMLUtils', function() {
```
HTMLUtils => HtmlUtils

andy-armstrong replied:
Fixed, and found a couple more too.

robrap commented:
Thanks @andy-armstrong.  This looks great to me.  I left 2 very minor comments.  :+1:  with those changes.  I'm happy with the current names.

That said, I have one possible additional enhancement.  Let's say we have the following example:
```
HtmlUtils.interpolate(
    gettext('You are enrolling a Rock & Roll star!')
)
```
This does exactly what we want, but the name `interpolate()` doesn't make a lot of sense because there is no interpolation needed.  We need to document what we'd expect to see used here.  One alternative that would work is:
```
_.escape(
    gettext('You are enrolling a Rock & Roll star!')
)
```
I'd propose we also have HtmlUtils provide the same so we can use it as an option (and as the best practice):
```
HtmlUtils.escape(
    gettext('You are enrolling a Rock & Roll star!')
)
```
I'd like your thoughts on this.  Do you agree, or would you prefer to keep `_.escape()` as the existing alternative because it is possibly well known and already documented as part of underscore?

Thanks.

andy-armstrong commented:
@robrap I really liked your suggestion of providing ``HtmlUtils.escape`` as a first class function. It isolates our best practices from Underscore (which we might drop at some point). More importantly, I found that it is useful to be able to pass HtmlSnippets to ``escape`` and have them pass through without being escaped again. In fact, having this method allowed me to remove the rather ugly internal method ``escapeInterpolateParameters`` with just ``_.mapObjects(parameters, escape)``. I'm also going to be using this method in edx-platform for exactly the use cases you describe, and in some of them the code won't know whether it has a raw string or an HtmlSnippet.

@robrap @dan-f @AlasdairSwan please re-review.

andy-armstrong commented:
@dan-f I've been thinking through the naming options for ``HtmlUtils.HTML``. I don't think we can use ``HtmlUtils.Html`` because that naming convention should indicate that ``Html`` is a class but it isn't. For example, JSHint complains about a class being constructed without using the ``new`` operator. ``HtmlUtils.html`` would be okay from a naming convention, but I find that it doesn't stand out enough. My current leaning is to go with @robrap's original proposal of ``HTML`` both because it stands out and because it mirrors what we're doing in Python.

Let me know if this works for you or not.

andy-armstrong commented:
@robrap @dan-f In working on the conversion of some of edx-platform, I'm starting to see a need for a non-HTML interpolate function. There are places where the code needs to create a text string and it is not going to be used as HTML. For example:

https://github.com/edx/edx-platform/blob/master/lms/djangoapps/teams/static/teams/js/views/edit_team_members.js#L50

I started to convert these functions over to use ``HtmlUtils.interpolate`` but the code didn't read correctly because it was jarring to see mention of HTML. However, I'd like to use the same syntax for consistency.

What would you think of also having ``StringUtils.interpolate`` which is a pure interpolation function that has nothing to do with HTML? Would that be confusing because there are two similar functions, or would it be less confusing because otherwise folks are relying on ``interpolate`` in the global namespace which has a somewhat confusing syntax?

I wonder if it would be possible to share logic between the two functions too, or whether that would just be too contrived.

robrap commented:
@andy-armstrong Thanks for your continued work on this.  I must have felt it because I woke up earlier thinking about this and finally got up to comment about some of the same things.  Also, thank you for actually putting your method to use to help discover these issues early.  :)

It struck me, as it did you, that `_.escape()` and `HtmlUtils.escape()` would be different in that the latter would recognize an `HtmlSnippet`, and that this difference is important and useful.  In the markupsafe library we use in Python, it uses this same convention.  In other words, it has an `escape()` method that won't double-escape if it is being passed `Markup`, which is the equivalent of an `HtmlSnippet`.  So again, I'm fine with the name `escape()`.

I'm on board with almost all existing names, except I think we may need to go with `HtmlUtils.interpolateHtml()` and `StringUtils.interpolate()`.  Especially for cases where the function is put into a variable, I think we'd want people to use `interpolate` and `interpolateHtml` to mean different things.  The only way to somewhat enforce this naming convention would be through the different function names to begin with.  I guess I don't know how often this is done, and how often one would see `interpolate()` (using the old name) separate from `HtmlUtils`?

Note that the linter will help enforce some of this.  For example, in Underscore.js templates we could dictate:
```
// Since "<%-" escapes, the linter would be happy with anything here except those methods found below
<%- ... %>

// Since "<%=" does not escape, the linter would be ok with the following:
<%= HtmlUtils.interpolateHtml(...) %>
<%= interpolateHtml(...) %>
<%= HtmlUtils.escape(...) %>
<%= escape(...) %>
<%= _.escape(...) %>

// The linter would NOT be ok with the following:
<%= interpolate(...) %>
 
// If we wanted to stick with "HtmlUtils.interpolate", the linter would instead be ok with the following:
<%= HtmlUtils.interpolate(...) %>
```

So, the linter can help somewhat in an Underscore.js Template file, but it is going to have a much more difficult time in a plain old JavaScript file.  So, I'll start with a vote for `HtmlUtils.interpolateHtml()`.

andy-armstrong commented:
@robrap Thanks for the great suggestions. I've introduced ``StringUtils.interpolate`` and renamed the original method to ``HtmlUtils.interpolateHtml``. It turned out to be very simple to implement the latter in terms of the former too.

One thing that we should discuss is how developers should handle escaping and/or interpolation in Underscore templates. The challenge is that none of these helpers are available in the global namespace, so it isn't straightforward to access them. The simplest thing to do would be to have the client code pass ``HtmlUtils`` through to the template. However, this would only be a best practice and would require every developer to think of it. We could have them use ``require`` directly in the template, but I don't know whether RequireJS Optimizer will be able to handle that appropriately (we would have to try). @AlasdairSwan and @dan-f, do you have any suggestions?

robrap commented:
@andy-armstrong Thanks.  I'm open to whatever you think will work in practicality.  I know @dan-f doesn't like the idea of extending underscore, but clearly enough people wanted this that it is available through the library using [mixin](http://underscorejs.org/#mixin).

Would something like `_.HtmlUtils` or `_.edX.HtmlUtils` make sense as a simple name-spaced solution that provides easy access?  I'm just throwing out ideas.  I do want it to be relatively simple for developers to use what they need.  I assume that this is just the first of many things you are going to want to easily make available.

dan-f commented:
:+1: on `HtmlUtils.HTML` as the factory function to return html safe text.
:+1: on `HtmlUtils.escape` as a function which smartly escapes and never double-escapes.

As for `StringUtils.interpolate` - I'd argue that we don't need. 90% of the time that we interpolate a string, we are preparing a string for the user to see. So it's going to end up in HTML. Hence when we do the interpolation, we want to mark HTML-safe strings with the `HTML` factory. So your example would become:
```javascript
var HtmlUtils = require('HtmlUtils'),
    interpolate = HtmlUtils.interpolate,
    HTML = HtmlUtils.HTML;

_.each(this.model.get('membership'), function(membership) {
    dateJoined = interpolate(
        // Translators: 'date' is a placeholder for a fuzzy, relative timestamp (see: https://github.com/rmm5t/jquery-timeago)
        gettext("Joined %(date)s"),
        {date: HTML(self.dateTemplate({date: membership.date_joined}))}
    );

```

dan-f commented on src/js/utils/html-utils.js:
```diff
@@ -0,0 +1,114 @@
+        }
+        HtmlSnippet.prototype.toString = function() {
+            return this.text;
+        };
```
It could also be useful to implement `valueOf`: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/valueOf

andy-armstrong replied:
Great idea. Done.

dan-f commented:
@andy-armstrong - I've started following a pattern in which string interpolation/preparation happens exclusively in JS, not in the template, which avoids this issue. I think that to me, it counts as view logic; the logic where we decide _what_ to present to the user. I like the template to be lightweight and simply determine _how_ that information is presented to the user.

andy-armstrong commented:
@dan-f The need for ``StringUtils.interpolate`` is when you are building up a string for some other purpose than HTML, e.g. to post it to AJAX, or to write it to the console, or any host of other reasons. It also might be an intermediate text string that will eventually be placed into HTML but might not be. I found that using ``HtmlUtils.interpolateHtml`` for such code just read as incorrect, and I think it would be functionally incorrect too if it was then used in a non-HTML context.

If it would help, you can see a number of such usages as I'm cleaning up teams to use these new functions:

https://github.com/edx/edx-platform/pull/11799/files

andy-armstrong commented:
@robrap @dan-f @AlasdairSwan I've pushed up another change which introduces a new ``HtmlUtils.template`` function. This wraps ``_.template`` but passes in the variable ``HtmlUtils`` so that all templates have access to these utilities. In addition, it returns an ``HtmlSnippet`` so that any consumers know that the result has already been escaped appropriately.

Please re-review at your convenience.

andy-armstrong commented:
@dan-f I spoke with @robrap about your suggestion of doing the interpolation and escaping in JavaScript. His recommendation is that the code always escapes only when the context is known, so his preference is to escape in the template. His concern with your approach is that all Underscore templates will have to use ``<%=`` and assume that the parameters passed in were appropriately escaped. This will also make it impossible for his linter to detect XSS problems.

We decided to add ``HtmlUtils.template`` to make it easy for templates to do this. I'm going to try using it and see whether it helps or not.

andy-armstrong commented:
@bjacobel FYI this is the XSS PR that I'm currently working on.

AlasdairSwan commented:
Looks good @andy-armstrong :+1:

robrap commented:
Based on the discussion, the linter would be ok with `<%=` (which does not escape), when used with the following:
```
<%= HtmlUtils.interpolateHtml(...) %>
<%= HtmlUtils.escape(...) %>
<%= HtmlUtils.HTML(...) %>
<%= _.escape(...) %>
```

robrap commented on src/js/utils/specs/html-utils-spec.js:
```diff
@@ -0,0 +1,106 @@
+                });
+            });
+
+            describe('template', function() {
```
Can you add a test for the fact that it is returning an HtmlSnippet that should not get escaped again if wrapped with another call like HtmlUtils.escape()?

andy-armstrong replied:
Added.

robrap commented:
:+1: Nice job.  Thanks!

dan-f commented:
@andy-armstrong and @robrap: I agree that you should always escape as close to the context/template as possible. You can still do so in JS:

```javascript
var template = _.template(templateName);
var message = interpolate(...);
this.$el.html(template({message: message}));
```

andy-armstrong commented:
@dan-f The problem with this is that the template has to now assume that the parameter it receives is escaped. So the template will have to use:

    <p><%= message %></p>

and that is precisely what @robrap's new linter will flag as being a potential issue. We want templates to default to safe (i.e. use ``<%-``) except in a few limited situations.

One way to combine your approach and still be safe is to use ``HtmlUtils.interpolateHtml`` in the JavaScript and then do the following in the template:

    <p><%= HtmlUtils.escape(message) %></p>

This template will then escape all raw strings, but will allow previously escaped HtmlSnippets to come through as is. This also won't cause an exception with the linter as ``HtmlUtils.escape`` is one of the known-to-be-safe patterns. Does that seem like a good compromise?

@robrap and I are going to be updating the best practice documentation to make all of this explicit for next week's XSS marathon.

cahrens commented on src/js/utils/string-utils.js:
```diff
@@ -0,0 +1,59 @@
+        /**
+         * Returns a string created by interpolating the provided parameters.
+         *
+         * The HTML text is provided as a tokenized format string where parameters
```
@andy-armstrong it looks like you forgot to update the documentation after this method moved out of HTMLUtils? This isn't HTML text, it's just a string, correct?

cahrens commented on src/js/utils/string-utils.js:
```diff
@@ -0,0 +1,59 @@
+         * replaced by the parameter value of the same name.
+         *
+         * Parameter values will be rendered using their toString methods and then
+         * HTML-escaped. The only exception is that instances of the class HTML
```
Is this HTML-escaping at all? I thought this method is just about interpolation.

cahrens commented on src/js/utils/html-utils.js:
```diff
@@ -0,0 +1,145 @@
+         * Returns a function that renders an Underscore template as an HTML snippet.
+         *
+         * Note: This helper function makes the following context parameters
+         * available to the template in addition to those passed in:
```
Are there supposed to be multiple parameters? I'm having trouble understanding the documentation.
